app/__init__.py

```diff
@@ -203,6 +203,11 @@ def exception(error):
             # error.code is set for our exception types.
             return jsonify(result='error', message=str(error)), 500
 
+    @app.errorhandler(400)
+    def bad_request(error):
+        app.logger.warning(str(error))
```

Comment:
So in theory, I'm OK with this being a warning level log.

However, in practice, I think we should keep this as an error level log. This is because the only people calling template preview should be us, and if we are getting 400s then it means something is not working as expected. A single one of these should cause us to investigate/log a bug about it. If we downgrade this to a warning, there is little chance we'd see it and little chance we'd get reminded of the fact there is a bug (or if there were similar bugs introduced in the future).

There could be an argument to downgrade this to a warning and instead raise the error at the client end calling template preview instead (I assume that isn't already happening).

Note, `error` is appropriate rather than `exception` as `exception` will give you the stack trace too which I don't think is relevant for us in this case.

Also note, I may be missing or misunderstanding something here with this particular change. I think the intention is to stop us getting emails each time we get a 4xx on template preview but do correct me where I'm wrong!

Reply:
Fair point. I assumed the missing handler for `400` was an omission, but perhaps the `Exception` handler is meant to pick it up. I think having an explicit handler is clearer though.

I've added a commit to fix this here: https://github.com/alphagov/notifications-template-preview/pull/637/commits/ae8dcde0e2f450456b7e83b0001a377d25cf6673

## So why are we getting this error?

I did some more investigation and found a way to reproduce it:

1. Create a template with placeholders that (just) wraps onto a second page.
2. Prepare to send a letter but use short values so the letter is just one page.
3. Instead of sending, go "back" a couple of steps (to the address field).
4. See that you get a "400 Bad Request: Letter does not have a page 2" error.

### Screenshots

| Step 1 | Step 2 | Step 3 |
| - | - | - |
| ![screencapture-notifications-service-gov-uk-services-3b525922-bdf9-4589-9d7b-dfe969f57805-send-7084e849-c427-41af-8942-f6224befed30-one-off-address-2022-03-01-11_57_13](https://user-images.githubusercontent.com/9029009/156165136-e505b59c-431f-4781-b226-99972981756f.png) |  ![screencapture-notifications-service-gov-uk-services-3b525922-bdf9-4589-9d7b-dfe969f57805-send-7084e849-c427-41af-8942-f6224befed30-one-off-step-7-2022-03-01-11_56_45](https://user-images.githubusercontent.com/9029009/156165156-8ce3d268-f724-4eb7-938a-2222e6b957bb.png) | ![screencapture-notifications-service-gov-uk-services-3b525922-bdf9-4589-9d7b-dfe969f57805-template-7084e849-c427-41af-8942-f6224befed30-notification-check-2022-03-01-11_57_00](https://user-images.githubusercontent.com/9029009/156165166-1b8caf7b-8ba9-4091-8fb3-3eacc07e3254.png) |

## What to do next

This PR could still go in (amended as above).

I can write this up as a low impact/effort bug card.

How does that sound?

Reply:
Yes please on writing a bug card. Great work on debugging this, I don't think anyone has been able to work this one for ages what was causing it.

I agree on the low impact. I haven't given it enough thought to know how much effort it would be to fix but I'm happy to go by your suggestion of low effort too.

Reply:
Also thanks for taking on board the log level stuff I raised initially!